### PR TITLE
[Model] Recurrent Relational Network (RRN) on sudoku

### PR DESCRIPTION
**Pull request by HuXiangkun**
## Description
Recurrent Relational Networks (RRN) model for solving Sudoku puzzles. The RRN module is a general framework for building RRNs for tasks which need multi-step reasoning.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/rrn/sudoku_data.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    def __getitem__(self, index):
+        return tuple(d[index] for d in self.lists_of_data)
+
+    def __len__(self):
+        return len(self.lists_of_data[0])
+
+
+def _get_sudoku_dataset(segment='train'):
+    assert segment in ['train', 'valid', 'test']
+    url = "https://www.dropbox.com/s/rp3hbjs91xiqdgc/sudoku-hard.zip?dl=1"
```
Comment (yzh119) [inline examples/pytorch/rrn/sudoku_data.py]: Please upload the dataset to dgl S3 and use the new link.

File: examples/pytorch/rrn/sudoku_data.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    basic_graph = _basic_sudoku_graph()
+    sudoku_indices = np.arange(0, 81)
+    rows = sudoku_indices // 9
+    cols = sudoku_indices % 9
+
+    def collate_fn(batch):
+        graph_list = []
+        for q, a in batch:
+            # q = copy(a)
```
Comment (yzh119) [inline examples/pytorch/rrn/sudoku_data.py]: pls remove the redundants.

File: examples/pytorch/rrn/train_sudoku.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                    test_res.append(int(torch.equal(preds[i, :], target[i, :])))
+
+                test_loss.append(loss.cpu().detach().data)
+
+        test_acc = sum(test_res) / len(test_res)
+        print(f"Test loss {np.mean(test_loss)}, accuracy {test_acc}\n")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description='sudoku')
```
Comment (yzh119) [inline examples/pytorch/rrn/train_sudoku.py]:
```suggestion
    parser = argparse.ArgumentParser(description='Recurrent Relational Network on sudoku task.')
```

File: examples/pytorch/rrn/train_sudoku.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+def main(args):
+    device = torch.device('cuda' if torch.cuda.is_available() else "cpu")
+
+    if args.do_train:
+        if not os.path.exists(args.output_dir):
+            os.mkdir(args.output_dir)
+
+        logging.basicConfig(filename=os.path.join(args.output_dir, 'train.log'), filemode='w', level=logging.DEBUG)
+        # print = logging.info
```
Comment (yzh119) [inline examples/pytorch/rrn/train_sudoku.py]: Remove this.

File: examples/pytorch/rrn/sudoku_data.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            return [(q, a) for q, a in reader]
+
+    data = read_csv(segment + '.csv')
+
+    def encode(samples):
+        def parse(x):
+            return list(map(int, list(x)))
+
+        encoded = [(parse(q), parse(a)) for q, a in samples]
+        # q, a = zip(*encoded)
```
Comment (yzh119) [inline examples/pytorch/rrn/sudoku_data.py]: Remove it.

File: examples/pytorch/rrn/rrn.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import torch
```
Comment (yzh119) [inline examples/pytorch/rrn/rrn.py]: Please provide header comments:
```suggestion
"""Recurrent Relational Network(RRN) module

References:
- Recurrent Relational Networks
- Paper: https://arxiv.org/abs/1711.08028
- Code: https://github.com/rasmusbergpalm/recurrent-relational-networks
"""
import torch
```

File: examples/pytorch/rrn/train_sudoku.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    batch_size = preds.size(0)
+    right_count = 0
+    for i in range(batch_size):
+        if torch.all(torch.eq(preds[i], target[i])).__bool__():
+            right_count += 1
+    return right_count / batch_size
+
+
+def main(args):
+    device = torch.device('cuda' if torch.cuda.is_available() else "cpu")
```
Comment (yzh119) [inline examples/pytorch/rrn/train_sudoku.py]: We should let user decide whether to use cpu/gpu with given id.
Please refer to https://github.com/dmlc/dgl/blob/master/examples/pytorch/gcn/train.py.

**Comment by jermainewang** [comment]
Cool work! I wonder how can we make a cool visualization?

**Pull request merged**